### ISSUE TITLE
Issue with packages not using src or lib folder

### ISSUE THREAD

**Issue by Nyholm** [Roave/BetterReflection#551]
Im running Roave BC Check and get the following output

```
In IdentifierNotFound.php line 30:
                                                                               
  Roave\BetterReflection\Reflection\ReflectionClass "Psr\Log\Test\DummyTest"   
  could not be found in the located source     
```

It comes from [this class](https://github.com/php-fig/log/blob/master/Psr/Log/Test/LoggerInterfaceTest.php#L140) not following PSR-4 (or PSR-0). The Dummy test is just one example, but this practise is not rare. 

Im not sure if this is an issue we want to fix or create a work around to. Or maybe I should open an issue on https://github.com/Roave/BackwardCompatibilityCheck

The [current status](https://github.com/async-aws/sqs/runs/466652035#step:4:44) is that I cannot use the BC check on projects that depends on `psr/log`.

**Comment by Ocramius** [comment]
Not an issue of either project, TBH.

Few issues here:

 1. tests should not be in vendor in first place
 2. If packages do not rely on composer autoloading, there must be a custom source locator, or a classmap autoloader

I'd escalate upstream before starting to build workarounds.

**Comment by asgrim** [comment]
> I'd escalate upstream before starting to build workarounds.

Whilst empirically, I completely agree with this, I can also see this:

> I cannot use the BC check on projects that depends on `psr/log`

being a problem... that's gonna be a LOT of projects. We're more likely to be able to move quicker and ship a release than `psr/log` is potentially. I'd need to look into exactly where the issue comes from (I tried looking earlier but there's a lot of complexity of locators being used for BC check :worried: so I stopped in favour of paid work).

Without looking deeper, I'm not sure what the best approach is. If we have to fudge stuff and make a hack in either BR or BC Checker just for this case, I'd rather push for upstream fix in `psr/log`. If we can make an easy and elegant workaround, I don't think it'd be _too_ bad.... would it....?

/me waits for raptors

**Comment by Ocramius** [comment]
Sending a PR upstream should be exactly like with any other project: psr doesn't really make this any special

**Comment by Nyholm** [comment]
I submitted a PR to https://github.com/php-fig/log. Make a comment if you want. 

I believe it is more project having this problem too. It would be great if there were a workaround. 

An example workaround could be a config file where I can define problems to ignore.

**Comment by Ocramius** [comment]
This failure specifically should be suppressed: https://github.com/async-aws/sqs/runs/466652035#step:4:41

If you can produce a trace, we can see what's responsible for a hard crash, but an error should indeed be raised by the tool nonetheless.

**Comment by Nyholm** [comment]
Here is a debug stacktrace generated with `debug_print_backtrace`.

```
#0  Roave\BetterReflection\Reflector\Exception\IdentifierNotFound::fromIdentifier(Roave\BetterReflection\Identifier\Identifier Object ([Roave\BetterReflection\Identifier\Identifiername] => Psr\Log\Test\DummyTest,[Roave\BetterReflection\Identifier\Identifiertype] => Roave\BetterReflection\Identifier\IdentifierType Object ([Roave\BetterReflection\Identifier\IdentifierTypename] => Roave\BetterReflection\Reflection\ReflectionClass))) called at [/Users/tobias/Workspace/PHPStorm/aws-core/vendor/roave/better-reflection/src/Reflector/ClassReflector.php:39]
#1  Roave\BetterReflection\Reflector\ClassReflector->reflect(Psr\Log\Test\DummyTest) called at [/Users/tobias/Workspace/PHPStorm/aws-core/vendor/roave/backward-compatibility-check/src/CompareClasses.php:79]
#2  Roave\BackwardCompatibility\CompareClasses->makeSymbolsIterator(Array ([0] => AsyncAws\Core\Signer\Signer,[1] => AsyncAws\Core\Signer\SignerV4,[4] => AsyncAws\Core\Test\Http\SimpleMockedResponse,[5] => AsyncAws\Core\Test\TestCase,[13] => AsyncAws\Core\Tests\Unit\Result\GetCallerIdentityResponseTest,[14] => AsyncAws\Core\Tests\Unit\Result\AssumeRoleResponseTest,[15] => AsyncAws\Core\Tests\Unit\StsClientTest,[16] => AsyncAws\Core\Tests\Unit\Input\AssumeRoleRequestTest,[17] => AsyncAws\Core\Tests\Unit\Input\GetCallerIdentityRequestTest,[18] => AsyncAws\Core\Tests\Unit\Stream\ResourceStreamTest,[19] => AsyncAws\Core\Tests\Unit\Stream\StringStreamTest,[20] => AsyncAws\Core\Tests\Unit\Stream\IterableStreamTest,[21] => AsyncAws\Core\Tests\Unit\Stream\FixedSizeStreamTest,[22] => AsyncAws\Core\Tests\Unit\Stream\CallableStreamTest,[23] => AsyncAws\Core\Tests\Unit\ResultTest,[24] => AsyncAws\Core\Tests\Unit\Exception\Http\ClientExceptionTest,[25] => AsyncAws\Core\Tests\Integration\StsClientTest,[26] => AsyncAws\Core\Result,[27] => AsyncAws\Core\AbstractApi,[28] => AsyncAws\Core\Configuration,[29] => AsyncAws\Core\Waiter,[30] => AsyncAws\Core\StreamableBody,[31] => AsyncAws\Core\StreamableBodyInterface,[32] => AsyncAws\Core\AwsClientFactory,[33] => AsyncAws\Core\Exception\MissingDependency,[34] => AsyncAws\Core\Exception\LogicException,[35] => AsyncAws\Core\Exception\InvalidArgument,[36] => AsyncAws\Core\Exception\RuntimeException,[37] => AsyncAws\Core\Exception\Http\ServerException,[38] => AsyncAws\Core\Exception\Http\HttpException,[39] => AsyncAws\Core\Exception\Http\ClientException,[41] => AsyncAws\Core\Exception\Http\RedirectionException,[42] => AsyncAws\Core\Exception\Http\NetworkException,[43] => AsyncAws\Core\Exception\Exception,[44] => Psr\Container\NotFoundExceptionInterface,[45] => Psr\Container\ContainerInterface,[46] => Psr\Container\ContainerExceptionInterface,[47] => Psr\Log\LoggerAwareInterface,[48] => Psr\Log\LogLevel,[49] => Psr\Log\Test\LoggerInterfaceTest,[50] => Psr\Log\Test\DummyTest,[51] => Psr\Log\Test\TestLogger,[52] => Psr\Log\LoggerAwareTrait,[53] => Psr\Log\InvalidArgumentException,[54] => Psr\Log\NullLogger,[55] => Psr\Log\LoggerInterface,[56] => Psr\Log\LoggerTrait,[57] => Psr\Log\AbstractLogger,[59] => JsonException,[60] => Symfony\Component\HttpClient\DataCollector\HttpClientDataCollector,[61] => Symfony\Component\HttpClient\HttpClientTrait,[62] => Symfony\Component\HttpClient\HttpOptions,[64] => Symfony\Component\HttpClient\Response\MockResponse,[65] => Symfony\Component\HttpClient\Response\StreamWrapper,[68] => Symfony\Component\HttpClient\Response\ResponseStream,[75] => Symfony\Component\HttpClient\HttplugClient,[76] => Symfony\Component\HttpClient\NativeHttpClient,[77] => Symfony\Component\HttpClient\ScopingHttpClient,[78] => Symfony\Component\HttpClient\MockHttpClient,[85] => Symfony\Component\HttpClient\HttpClient,[86] => Symfony\Component\HttpClient\CachingHttpClient,[87] => Symfony\Component\HttpClient\CurlHttpClient,[88] => Symfony\Component\HttpClient\TraceableHttpClient,[89] => Symfony\Component\HttpClient\DependencyInjection\HttpClientPass,[90] => Symfony\Component\HttpClient\Exception\ServerException,[91] => Symfony\Component\HttpClient\Exception\ClientException,[93] => Symfony\Component\HttpClient\Exception\InvalidArgumentException,[94] => Symfony\Component\HttpClient\Exception\RedirectionException,[95] => Symfony\Component\HttpClient\Exception\TransportException,[96] => Symfony\Component\HttpClient\Exception\JsonException,[97] => Symfony\Component\HttpClient\Psr18Client,[100] => Symfony\Contracts\Service\ServiceSubscriberInterface,[101] => Symfony\Contracts\Service\Test\ServiceLocatorTest,[103] => Symfony\Contracts\Service\ResetInterface,[104] => Symfony\Contracts\Service\ServiceProviderInterface,[105] => Symfony\Contracts\Service\ServiceSubscriberTrait,[106] => Symfony\Contracts\Service\ServiceLocatorTrait,[109] => Symfony\Contracts\HttpClient\ResponseStreamInterface,[110] => Symfony\Contracts\HttpClient\Test\TestHttpServer,[111] => Symfony\Contracts\HttpClient\Test\HttpClientTestCase,[112] => Symfony\Contracts\HttpClient\HttpClientInterface,[113] => Symfony\Contracts\HttpClient\ChunkInterface,[114] => Symfony\Contracts\HttpClient\ResponseInterface,[115] => Symfony\Contracts\HttpClient\Exception\ExceptionInterface,[116] => Symfony\Contracts\HttpClient\Exception\ServerExceptionInterface,[117] => Symfony\Contracts\HttpClient\Exception\RedirectionExceptionInterface,[118] => Symfony\Contracts\HttpClient\Exception\HttpExceptionInterface,[119] => Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface,[120] => Symfony\Contracts\HttpClient\Exception\DecodingExceptionInterface,[121] => Symfony\Contracts\HttpClient\Exception\ClientExceptionInterface,[122] => AsyncAws\Core\Sts\StsClient,[123] => AsyncAws\Core\Sts\Result\AssumedRoleUser,[124] => AsyncAws\Core\Sts\Result\Credentials,[125] => AsyncAws\Core\Sts\Result\AssumeRoleResponse,[126] => AsyncAws\Core\Sts\Result\GetCallerIdentityResponse,[127] => AsyncAws\Core\Sts\Input\PolicyDescriptorType,[128] => AsyncAws\Core\Sts\Input\Tag,[129] => AsyncAws\Core\Sts\Input\GetCallerIdentityRequest,[130] => AsyncAws\Core\Sts\Input\AssumeRoleRequest,[131] => AsyncAws\Core\Credentials\CacheProvider,[132] => AsyncAws\Core\Credentials\Credentials,[133] => AsyncAws\Core\Credentials\ChainProvider,[134] => AsyncAws\Core\Credentials\CredentialProvider,[135] => AsyncAws\Core\Credentials\NullProvider,[136] => AsyncAws\Core\Credentials\ConfigurationProvider,[137] => AsyncAws\Core\Credentials\IniFileProvider,[138] => AsyncAws\Core\Credentials\InstanceProvider), Roave\BetterReflection\Reflector\ClassReflector Object ([Roave\BetterReflection\Reflector\ClassReflectorsourceLocator] => Roave\BetterReflection\SourceLocator\Type\MemoizingSourceLocator Object ([Roave\BetterReflection\SourceLocator\Type\MemoizingSourceLocatorwrappedSourceLocator] => Roave\BetterReflection\SourceLocator\Type\AggregateSourceLocator Object ([Roave\BetterReflection\SourceLocator\Type\AggregateSourceLocatorsourceLocators] => Array ([0] => Roave\BetterReflection\SourceLocator\Type\AggregateSourceLocator Object ([Roave\BetterReflection\SourceLocator\Type\AggregateSourceLocatorsourceLocators] => Array ([0] => Roave\BetterReflection\SourceLocator\Type\Composer\PsrAutoloaderLocator Object ([Roave\BetterReflection\SourceLocator\Type\Composer\PsrAutoloaderLocatormapping] => Roave\BetterReflection\SourceLocator\Type\Composer\Psr\Psr4Mapping Object ([Roave\BetterReflection\SourceLocator\Type\Composer\Psr\Psr4Mappingmappings] => Array ([AsyncAws\Core\] => Array ([0] => /private/var/folders/hw/xb7mf0_5453d0bqt970t1mkh0000gn/T/api-compare-7356e1b6562ee8c95f176c11c2483aec66db81bf_5e57be00c2798))),[Roave\BetterReflection\SourceLocator\Type\Composer\PsrAutoloaderLocatorastLocator] => Roave\BetterReflection\SourceLocator\Ast\Locator Object ([Roave\BetterReflection\SourceLocator\Ast\LocatorfindReflectionsInTree] => Roave\BetterReflection\SourceLocator\Ast\FindReflectionsInTree Object ([Roave\BetterReflection\SourceLocator\Ast\FindReflectionsInTreeastConversionStrategy] => Roave\BetterReflection\SourceLocator\Ast\Strategy\NodeToReflection Object (),[Roave\BetterReflection\SourceLocator\Ast\FindReflectionsInTreefunctionReflector] => ,[Roave\BetterReflection\SourceLocator\Ast\FindReflectionsInTreefunctionReflectorGetter] => Closure Object ()),[Roave\BetterReflection\SourceLocator\Ast\Locatorparser] => Roave\BetterReflection\SourceLocator\Ast\Parser\MemoizingParser Object ([Roave\BetterReflection\SourceLocator\Ast\Parser\MemoizingParsersourceHashToAst] => Array ([a708225d1a63ace05119e35e5dac546873ccbe78d623743d4c6d958a65e39d45:279] => Array ([0] => PhpParser\Node\Stmt\Namespace_ Object ([name] => PhpParser\Node\Name Object ([parts] => Array ([0] => AsyncAws,[1] => Core,[2] => Signer),[*attributes] => Array ([startLine] => 3,[startFilePos] => 17,[endLine] => 3,[endFilePos] => 36)),[stmts] => Array ([0] => PhpParser\Node\Stmt\Use_ Object ([type] => 1,[uses] => Array ([0] => PhpParser\Node\Stmt\UseUse Object ([type] => 0,[name] => PhpParser\Node\Name Object ([parts] => Array ([0] => AsyncAws,[1] => Core,[2] => Credentials,[3] => Credentials),[*attributes] => Array ([startLine] => 5,[startFilePos] => 44,[endLine] => 5,[endFilePos] => 80)),[alias] => ,[*attributes] => Array ([startLine] => 5,[startFilePos] => 44,[endLine] => 5,[endFilePos] => 80))),[*attributes] => Array ([startLine] => 5,[startFilePos] => 40,[endLine] => 5,[endFilePos] => 81)),[1] => PhpParser\Node\Stmt\Interface_ Object ([extends] => Array (),[name] => PhpParser\Node\Identifier Object ([name] => Signer,[*attributes] => Array ([startLine] => 12,[startFilePos] => 191,[endLine] => 12,[endFilePos] => 196)),[stmts] => Array ([0] => PhpParser\Node\Stmt\ClassMethod Object ([flags] => 1,[byRef] => ,[name] => PhpParser\Node\Identifier Object ([name] => sign,[*attributes] => Array ([startLine] => 14,[startFilePos] => 220,[endLine] => 14,[endFilePos] => 223)),[params] => Array ([0] => PhpParser\Node\Param Object ([type] => PhpParser\Node\Name\FullyQualified Object ([parts] => Array ([0] => AsyncAws,[1] => Core,[2] => Signer,[3] => Request),[*attributes] => Array ([startLine] => 14,[startFilePos] => 225,[endLine] => 14,[endFilePos] => 231)),[byRef] => ,[variadic] => ,[var] => PhpParser\Node\Expr\Variable Object ([name] => request,[*attributes] => Array ([startLine] => 14,[startFilePos] => 233,[endLine] => 14,[endFilePos] => 240)),[default] => ,[*attributes] => Array ([startLine] => 14,[startFilePos] => 225,[endLine] => 14,[endFilePos] => 240),[isOptional] => ),[1] => PhpParser\Node\Param Object ([type] => PhpParser\Node\NullableType Object ([type] => PhpParser\Node\Name\FullyQualified Object ([parts] => Array ([0] => AsyncAws,[1] => Core,[2] => Credentials,[3] => Credentials),[*attributes] => Array ([startLine] => 14,[startFilePos] => 244,[endLine] => 14,[endFilePos] => 254)),[*attributes] => Array ([startLine] => 14,[startFilePos] => 243,[endLine] => 14,[endFilePos] => 254)),[byRef] => ,[variadic] => ,[var] => PhpParser\Node\Expr\Variable Object ([name] => credentials,[*attributes] => Array ([startLine] => 14,[startFilePos] => 256,[endLine] => 14,[endFilePos] => 267)),[default] => ,[*attributes] => Array ([startLine] => 14,[startFilePos] => 243,[endLine] => 14,[endFilePos] => 267),[isOptional] => )),[returnType] => PhpParser\Node\Identifier Object ([name] => void,[*attributes] => Array ([startLine] => 14,[startFilePos] => 271,[endLine] => 14,[endFilePos] => 274)),[stmts] => ,[*attributes] => Array ([startLine] => 14,[startFilePos] => 204,[endLine] => 14,[endFilePos] => 275))),[*attributes] => Array ([startLine] => 12,[startFilePos] => 181,[comments] => Array ([0] => PhpParser\Comment\Doc Object ([*text] => /**

```


> This failure specifically should be suppressed

True.

**Comment by Ocramius** [comment]
Ok, so what has happened here (weird, but it happened), is that `ClassReflector#getAllClasses()` found `Psr\Log\Test\DummyTest`: https://github.com/Roave/BackwardCompatibilityCheck/blob/3dd77835db29c46e571c41acedf733766015b871/src/CompareClasses.php#L53

After that, a direct lookup via `ClassReflector#reflect()` failed: https://github.com/Roave/BackwardCompatibilityCheck/blob/3dd77835db29c46e571c41acedf733766015b871/src/CompareClasses.php#L78-L79

This is because the first `ClassReflector` and the second one have different `SourceLocator`s attached: https://github.com/Roave/BackwardCompatibilityCheck/blob/3dd77835db29c46e571c41acedf733766015b871/src/Command/AssertBackwardsCompatible.php#L163-L170

Now the problem is that `Psr\Log\Test\DummyTest` is picked up as a **source** class.

In practice, this bit should never find `Psr\Log\Test\DummyTest`, as it isn't part of the sources: https://github.com/Roave/BackwardCompatibilityCheck/blob/3dd77835db29c46e571c41acedf733766015b871/src/Command/AssertBackwardsCompatible.php#L163-L166

What we could need is a small reproducer (repo would suffice, since it really needs some debugging work) that makes the referenced code above list `Psr\Log\Test\DummyTest` under sources (bug).

**Comment by Nyholm** [comment]
Thank you!

Have a look at this repo: https://github.com/Nyholm/better-reflection-debug

**Comment by Nyholm** [comment]
I found another thing. 

See this repo: https://github.com/Nyholm/better-reflection-debug-2

I dont have any issues here. This repo is using a `src` folder.

**Comment by Ocramius** [comment]
Hmm, I'd say that autoloading is broken: https://github.com/Nyholm/better-reflection-debug/blob/88a2419c4cfeaf996e56f97b3050ce8586c64fe7/composer.json#L15-L17

This will consider anything under `"./"` as source file, including `"vendor/"` (which obviously isn't source)

**Comment by Nyholm** [comment]
Hm... That is not 100% correct. 
Composer will expect to find all `Debug\` classes in the project root. So may it will look in the vendor folder if I try to load `Debug\vendor\psr\log\Psr\Log\NullLogger`

It does not say "autolaod everything here. 

I tested by [adding my vendor folder](https://github.com/Nyholm/better-reflection-debug/commit/2614541822a56f40af6dffca15c6e4c1abae74a6) to git, then added "eclude from clasmap" and ran "dump-autoload". https://github.com/Nyholm/better-reflection-debug/commit/9b512a635e5741bae1cf30bfabfe12b7d533b843

**Comment by Ocramius** [comment]
In theory, `exclude-from-classmap` was supposed to eliminate autoloading for `dev-only` paths, but Symfony pretty much adopted it rather than fixing their directory structure.

Given that composer supports it (https://github.com/composer/composer/pull/1607), and there's not much hope for folks to fix this broken convention in their downstream repos (would require deprecation in composer first, and the feature is objectively useful), I think it will need to be added as new feature.

`"exclude-from-classmap"` is not currently part of the source locator logic/factory for few reasons:

 * when I designed the factory, the `"exclude-from-classmap"` design was unbeknownst to me - discovered it very late in the design process
 * excessive complexity, as it is the only subtractive/negative mapping
 * is only in use in the symfony ecosystem (hopefully!)
 * if you use the widespread `src/` or `lib/` convention, it is eliminated already

It should be handled in https://github.com/Roave/BetterReflection/blob/40d8b7baac3c0157ea2b3433d5c5f5408327a4c7/src/SourceLocator/Type/Composer/Factory/MakeLocatorForComposerJson.php, could probably do with a filter iterator or such, but with much more complexity and testing.

 > It does not say "autoload everything here".

That's true, but the "source locator" identifies sources, not "autoloadable sources". "autoloadable" is a secondary concern, since it can wildly differ from repository to repository (heck, imagine stuff like `HTMLPurifier`, which ships their own autoloader), but if the convention of "sources" is not simple, then a simplistic factory won't suffice either.

**Comment by Nyholm** [comment]
I started doing [some work](https://github.com/Roave/BetterReflection/compare/master...Nyholm:filter-vendor) to exclude the vendor directly automatically, but Im not sure this is a correct beginning. I also see that there are some more issues with this attempt. 

I decided to "solve" my problem by listening to @Ocramius advice to fix the directory structure =)

I renamed the title so this issue can be more easily found. I will close this issue because the suggested solution is more appealing than trying to wedge in the "exclude" feature here. 

Thank you for the help.
